EIPS/eip-6732.md

```diff
@@ -0,0 +1,298 @@
+}
+```
+
+In addition, in order for a contract to be compliant with this ERC, it MUST abide by the following:
+
+- Implementers MAY enable token transfers to support functionality such as ticket reselling.
+- Implementers MUST NOT allow tokens to be transferred between addresses after they have been redeemed.
+- Implementers MUST allow token holders to redeem their tokens.
+- Implementers MUST NOT allow token issuers to redeem the tokens they have issued.
+- Implementers SHOULD allow token recipients to burn any token they receive.
+- Implementers MAY enable token issuers to burn the tokens they issued.
+
+**Smart contracts MUST implement all of the functions in the `IRedeemableTokenReceiver` interface to accept tokens being minted to them.**
+
+The **URI Storage extension** is OPTIONAL for compliant smart contracts. This allows contracts to associate a unique URI for each token ID.
+
+```solidity
+interface IRedeemableTokenURIStorage is IRedeemableToken {
+    /**
+     * @dev Emitted when the URI for token type `id` changes to `value`, if it is a non-programmatic URI.
+     */
+    event URI(uint256 indexed id, string value);
+
+    /**
+     * @dev Returns the Uniform Resource Identifier (URI) for `id` token.
+     */
+    function tokenURI(uint256 id) external view returns (string memory);
+}
+```
+
+The **Expirable extension** is OPTIONAL for compliant smart contracts. This allows contracts to associate a unique expiry date for each token ID. Smart contracts implementing the `RedeemableTokenExpirable` extension MUST abide by the following:
+
+- Implementers MUST NOT allow tokens to be minted or redeemed if expired.
+
+```solidity
+interface IRedeemableTokenExpirable is IRedeemableToken {
+    /**
+     * @dev Sets the expiry date for the token of token type `id`.
+     */
+    function setExpiryDate(uint256 id, uint256 date) external;
+
+    /**
+     * @dev [Batched] version of {setExpiryDate}.
+     */
+    function setBatchExpiryDates(uint256[] memory ids, uint256[] memory dates) external;
+
+    /**
+     * @dev Returns the expiry date for the token of token type `id`.
+     */
+    function expiryDate(uint256 id) external view returns (uint256);
+
+    /**
+     * @dev Returns `true` or `false` depending on if the token of token type `id` has expired
+     * by comparing the expiry date with `block.timestamp`.
+     */
+    function isExpired(uint256 id) external view returns (bool);
+}
+```
+
+The **Supply extension** is OPTIONAL for compliant smart contracts. This allows contracts to associate a unique max supply for each token ID. Smart contracts implementing the `RedeemableTokenSupply` extension MUST abide by the following:
+
+- Implementers SHOULD NOT allow tokens to be minted if total supply exceeds max supply.
+- Implementers SHOULD increment total supply upon minting and decrement upon burning.
+- Implementers are RECOMMENDED to override the `_beforeMint` hook to increment total supply upon minting and decrement upon burning.
```

Comment:
The `_beforeMint` function is part of the reference implementation, not the specification itself. The Specification section should be understandable on its own without the other sections, while the reference implementation should help to clarify what's in the the Specification section.

```suggestion
```